scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -62,7 +60,41 @@ import scala.concurrent.{Future, Promise}
 import scala.io.Source
 import scala.reflect.ClassTag
 import scala.reflect.runtime.universe._
-import scala.util.Try
+import scala.util.{Failure, Success, Try}
+import scala.util.control.NonFatal
+
+/** Runner specific context. */
+trait RunnerContext {
+  def prepareOptions(options: PipelineOptions, artifacts: List[String]): Unit
+}
+
+private case object NoOpContext extends RunnerContext {
+  override def prepareOptions(options: PipelineOptions, artifacts: List[String]): Unit = Unit
+}
+
+/** Direct runner specific context. */
+private case object DirectContext extends RunnerContext {
+  override def prepareOptions(options: PipelineOptions, artifacts: List[String]): Unit = Unit
+}
+
+/** Companion object for [[RunnerContext]]. */
+private object RunnerContext {
+  private val mapping = Map(
+    "DirectRunner" -> DirectContext.getClass.getName,
+    "DataflowRunner" -> "com.spotify.scio.runners.dataflow.DataflowContext$")
+    .withDefaultValue(NoOpContext.getClass.getName)
```

Comment:
Would it be better to fail instead of some default `NoOpContext`?

Reply:
We could print a warning message, but should have some default behavior for unknown runners. Right now `RunnerContext` only injects Scio-REPL specific logic for `DataflowRunner`.

Reply:
Well if there is unknown runner wouldn't it fail anyway, so would it make sense to just fail asap (?)

Reply:
Unknown runner should work out of the box for the most case. We only use this to inject files to stage for DF runner. So we shouldn't fail them.

Reply:
@nevillelyh ah i see, that makes sense.